x/bsonx/bsoncore/tranquil_document.go

```diff
@@ -62,3 +66,91 @@ func (d Document) LookupCaseInsensitive(key string) (string, Value, error) {
 	}
 	return "", Value{}, ErrElementNotFound
 }
+
+func (d Document) LookupRecursiveReturnFirst(key string) (Value, error) {
+	length, rem, ok := ReadLength(d)
+	if !ok {
+		return Value{}, NewInsufficientBytesError(d, rem)
+	}
+
+	length -= 4
+
+	var elem Element
+	for length > 1 {
+		elem, rem, ok = ReadElement(rem)
+		length -= int32(len(elem))
+		if !ok {
+			return Value{}, NewInsufficientBytesError(d, rem)
+		}
+		// We use `KeyBytes` rather than `Key` to avoid a needless string alloc.
+		if string(elem.KeyBytes()) == key {
+			if elemVal, err := elem.ValueErr(); err != nil {
+				continue
+			} else {
+				return elemVal, nil
+			}
+		}
+		tt := bsontype.Type(elem[0])
+		switch tt {
+		case bsontype.EmbeddedDocument:
+			val, err := elem.Value().Document().LookupRecursiveReturnFirst(key)
+			if err == nil {
+				return val, err
+			}
+		case bsontype.Array:
+			val, err := elem.Value().Array().LookupRecursiveReturnFirst(key)
+			if err == nil {
+				return val, err
+			}
+		}
+	}
+	return Value{}, ErrElementNotFound
+}
+
+//LookupAllRecursive returns all instances of values keyed by key. These are not checked for uniqueness,
+// so callers must expect to occasionally receive duplicate values.
+func (d Document) LookupAllRecursive(key string) ([]Value, error) {
+	length, rem, ok := ReadLength(d)
+	if !ok {
+		return []Value{}, NewInsufficientBytesError(d, rem)
+	}
+
+	length -= 4
+
+	returnValues := make([]Value, 2)[0:0]
+	var elem Element
+	for length > 1 {
+		elem, rem, ok = ReadElement(rem)
+		length -= int32(len(elem))
+		if !ok {
+			return []Value{}, NewInsufficientBytesError(d, rem)
+		}
+		// We use `KeyBytes` rather than `Key` to avoid a needless string alloc.
+		if string(elem.KeyBytes()) == key {
+			if elemVal, err := elem.ValueErr(); err != nil {
+				continue
+			} else {
+				returnValues = append(returnValues, elemVal)
+			}
+		}
+		tt := bsontype.Type(elem[0])
+		switch tt {
+		case bsontype.EmbeddedDocument:
```

Comment:
I cannot remember if there is a 3rd kind of bsontype. If so, maybe add a "default" with some sort of warning or other appropriate handling.

Reply:
There are more than 3 bson types, but embedded document and array are the only two that can be recursed into